README.md
# refresh_state

<a href="https://pub.dev/packages/refresh_state"><img src="https://img.shields.io/pub/v/refresh_state.svg?label=refresh_state" alt="refresh_state version"></a>

<a href="https://developer.android.com" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-android-blue">
</a>
<a href="https://developer.apple.com/ios/" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-iOS-blue">
</a>
<a href="" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-Linux-blue">
</a>
<a href="" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-Mac-blue">
</a>
<a href="" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-web-blue">
</a>
<a href="" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-Windows-blue">
</a>
<a href="https://opensource.org/licenses/MIT"><img src="https://img.shields.io/badge/license-MIT-purple.svg" alt="MIT License"></a>

refresh_state provides state-management solution by eliminating redundant code, helps speed-up the development process

## Installation
Add refresh_state to your project's pubspec.yaml:

## Refresher provides builder & listener, has id which will be used to refresh the state.

### Refresher
    Refresher<int>(
                id: 'abc',
                initialData: 0,
                initialState: AppState.blue,
                listener: (state, data) {
                  if (kDebugMode) {
                    print('listening $state $data');
                  }
                },
                builder: (context, state, data) {
                  return Center(
                      child: CircleAvatar(
                    radius: 30,
                    backgroundColor: state == AppState.red
                        ? Colors.red
                        : state == AppState.green
                            ? Colors.green
                            : Colors.blue,
                    child: Text(
                      '$data',
                      style: const TextStyle(
                          fontSize: 18,
                          fontWeight: FontWeight.w500,
                          color: Colors.white),
                    ),
                  ));
                },
              )

### parameters


##### id:
id will be used to identify refresher, multiple Refresher can have same id, but if given then all the Refreshers with same id will be notified

#### initialData:
initialData will be initial-data given in builder.

#### initialState:
initialState will be initial-state given in builder.

#### listener:
listener will be called when refresh is called.

#### builder:
builder will be called when refresh is called.

#### child:
if only want to listen use child instead of builder,
builder and child can not be used together

## Call refresh function with ID from anywhere in the app to notify Refresher.

### refresh
    refresh('abc', (state) {
                  if (state == AppState.blue) {
                    return AppState.red;
                  } else if (state == AppState.red) {
                    return AppState.green;
                  } else {
                    return AppState.blue;
                  }
                });


### parameters

#### id:
id will be used to identify refresher, multiple Refresher can have same id, but if given then all the Refreshers with same id will be notified.

#### state-callback
state-callback will have old-state in argument, and new-state will be returned.

#### data-callback
data-callback will have old-data in argument, and new-data will be returned.


## Guideline for contributors
Contribution towards our repository is always welcome, we request contributors to create a pull request to the develop branch only.

# LICENSE!
refresh_state is [MIT-licensed](https://github.com/abrarmalekji12/refresh_state/blob/master/LICENSE "MIT-licensed").


CHANGELOG.md
## 0.0.2

* improved code.

## 0.0.1

* Initial release.


test/refresh_state_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:refresh_state/refresh_state.dart';

void main() {}


lib/refresh_state.dart
library refresh_state;

import 'package:flutter/material.dart';

final Map<String, List<_RefreshController>> _refresherMap = {};

/// Refresher provides optional builder and listener for the widget.
class Refresher<T> extends StatefulWidget {
  const Refresher(
      {Key? key,
      this.initialData,
      this.initialState,
      required this.id,
      this.builder,
      this.child,
      this.listener})
      : super(key: key);

  /// [id] use to differentiate refreshers, and will be used when we want to refresh.
  final String id;

  /// Initial data of the refresher.
  final T? initialData;

  /// Initial state of the refresher.
  final Enum? initialState;

  /// Builder for the refresher, if not provided then need to provide [child].
  final Widget Function(BuildContext, Enum?, T?)? builder;

  /// Child if [builder] is not provided.
  final Widget? child;

  /// [listener] for the refresher, it is optional
  final void Function(Enum?, T?)? listener;

  @override
  State<Refresher<T>> createState() => _RefresherState<T>();
}

class _RefresherState<T> extends State<Refresher<T>> {
  /// [state] is the state of the refresher.
  Enum? state;

  /// [data] is the data of the refresher.
  T? data;
  final _RefreshController<T> _refreshController = _RefreshController<T>();

  @override
  void initState() {
    super.initState();
    assert(widget.child != null || widget.builder != null);
    assert(widget.child == null || widget.builder == null);
    _refreshController.data = widget.initialData;
    _refreshController.state = widget.initialState;
    if (_refresherMap[widget.id] != null) {
      _refresherMap[widget.id]?.add(_refreshController);
    } else {
      _refresherMap[widget.id] = [_refreshController];
    }
    _refreshController.addListener(() {
      widget.listener?.call(_refreshController.state!, _refreshController.data);
    });
  }

  @override
  void dispose() {
    _refresherMap[widget.id]?.remove(_refreshController);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    if (widget.builder != null) {
      return AnimatedBuilder(
          animation: _refreshController,
          builder: (context, child) {
            return widget.builder!.call(
                context, _refreshController.state, _refreshController.data);
          });
    }
    return widget.child!;
  }
}

/// [refresh] is a function that will trigger builder and listener of the refresher.
void refresh<T>(String id, Enum? Function(Enum?) state,
    {T? Function(T?)? data}) {
  final refresherList = _refresherMap[id];
  assert(refresherList != null);
  for (final refresher in refresherList!) {
    refresher.refresh(state.call(refresher.state),
        data != null ? data.call(refresher.data) : refresher.data);
  }
}

class _RefreshController<T> extends ChangeNotifier {
  Enum? state;
  T? data;

  void refresh(Enum? state, T? data) {
    if (state != this.state || data != this.data) {
      this.state = state;
      this.data = data;
      notifyListeners();
    }
  }
}


example/lib/main.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:refresh_state/refresh_state.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

enum AppState { red, green, blue }

class _MyHomePageState extends State<MyHomePage> {
  void _incrementCounter() {
    refresh<int>('abc', (state) => state, data: (oldData) => oldData! + 1);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.spaceAround,
        children: [
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceAround,
            children: [
              Refresher<int>(
                id: 'abc',
                initialData: 0,
                initialState: AppState.blue,
                listener: (state, data) {
                  if (kDebugMode) {
                    print('listening $state $data');
                  }
                },
                builder: (context, state, data) {
                  return Center(
                      child: CircleAvatar(
                    radius: 30,
                    backgroundColor: state == AppState.red
                        ? Colors.red
                        : state == AppState.green
                            ? Colors.green
                            : Colors.blue,
                    child: Text(
                      '$data',
                      style: const TextStyle(
                          fontSize: 18,
                          fontWeight: FontWeight.w500,
                          color: Colors.white),
                    ),
                  ));
                },
              ),
              Refresher<int>(
                id: 'abc',
                initialData: 0,
                initialState: AppState.blue,
                builder: (context, state, data) {
                  return Center(
                      child: CircleAvatar(
                    radius: 30,
                    backgroundColor: state == AppState.red
                        ? Colors.red
                        : state == AppState.green
                            ? Colors.green
                            : Colors.blue,
                    child: Text(
                      '$data',
                      style: const TextStyle(
                          fontSize: 18,
                          fontWeight: FontWeight.w500,
                          color: Colors.white),
                    ),
                  ));
                },
              ),
            ],
          ),
          const SizedBox(height: 20),
          TextButton(
              style: ButtonStyle(
                  backgroundColor: MaterialStateProperty.all(Colors.blue)),
              onPressed: () {
                refresh('abc', (state) {
                  if (state == AppState.blue) {
                    return AppState.red;
                  } else if (state == AppState.red) {
                    return AppState.green;
                  } else {
                    return AppState.blue;
                  }
                });
              },
              child: const Text(
                'Switch Colors',
                style: TextStyle(color: Colors.white, fontSize: 18),
              )),
        ],
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
